server/main_test.go

```diff
@@ -0,0 +1,218 @@
+
+	tmpDir, err := os.MkdirTemp("", "msteams")
+	if err != nil {
+		return err
+	}
+	mt.Cleanup(func() {
+		os.RemoveAll(tmpDir)
+	})
+
+	// Execute from the temporary directory to avoid polluting the developer's working
+	// directory and simplify cleanup.
+	err = os.Chdir(tmpDir)
+	if err != nil {
+		return err
+	}
+
+	// Setup a custom MM_LOCALSOCKETPATH.
+	os.Setenv("MM_LOCALSOCKETPATH", path.Join(tmpDir, "mattermost_local.socket"))
+
+	// Create a test memory store and modify configuration appropriately
+	configStore := config.NewTestMemoryStore()
+	config := configStore.Get()
+	config.PluginSettings.Directory = model.NewString(path.Join(tmpDir, "plugins"))
+	config.PluginSettings.ClientDirectory = model.NewString(path.Join(tmpDir, "client"))
+	config.ServiceSettings.ListenAddress = model.NewString("localhost:0")
+	config.TeamSettings.MaxUsersPerTeam = model.NewInt(10000)
+	config.LocalizationSettings.SetDefaults()
+	config.SqlSettings = *storetest.MakeSqlSettings("postgres", false)
+	config.ServiceSettings.SiteURL = model.NewString("http://example.com/")
+	config.LogSettings.EnableConsole = model.NewBool(true)
+	config.LogSettings.EnableFile = model.NewBool(false)
+	config.LogSettings.ConsoleLevel = model.NewString("INFO")
+	config.ServiceSettings.EnableLocalMode = model.NewBool(true)
+	config.ServiceSettings.LocalModeSocketLocation = model.NewString(path.Join(tmpDir, "mattermost_local.socket"))
+	config.ServiceSettings.EnableDeveloper = model.NewBool(true)
+	config.ServiceSettings.EnableTesting = model.NewBool(true)
+	config.FileSettings.Directory = model.NewString(path.Join(tmpDir, "data"))
+
+	_, _, err = configStore.Set(config)
+	if err != nil {
+		return err
+	}
+
+	// Create a logger to override
+	testLogger, err := mlog.NewLogger()
+	if err != nil {
+		return err
+	}
+	testLogger.LockConfiguration()
+
+	// Initialize the server with app and api4 interfaces.
+	options := []app.Option{
+		app.ConfigStore(configStore),
+	}
+
+	server, err := app.NewServer(options...)
+	if err != nil {
+		return err
+	}
+
+	_, err = api4.Init(server)
+	if err != nil {
+		return err
+	}
+
+	err = server.Start()
+	if err != nil {
+		return err
+	}
+	mt.Cleanup(func() {
+		server.Shutdown()
+	})
+
+	ap := app.New(app.ServerConnector(server.Channels()))
+
+	// Setup the first user immediately.
+	username := model.NewId()
+	user := &model.User{
+		Email:         fmt.Sprintf("%s@example.com", username),
+		Username:      username,
+		Password:      "password",
+		EmailVerified: true,
+	}
+
+	_, appErr := ap.CreateUser(request.EmptyContext(testLogger), user)
+	if appErr != nil {
+		return appErr
+	}
+
+	// Export the site url for later use in tests.
+	os.Setenv("MM_SERVICESETTINGS_SITEURL", fmt.Sprintf("http://localhost:%v", server.ListenAddr.Port))
+
+	return nil
+}
+
+// TestMain is run before any tests within this package and helps setup a singleton Postgres and
+// Mattermost intance for use with tests.
+func TestMain(m *testing.M) {
```

Comment:
I opted for a singleton Mattermost + Postgres instance instead of spinning these up and down for each test. It's more interesting to start and stop the plugin as we need to (which is lightning fast).

Reply:
Let's see how many tests break due to missing cleanup in other tests.

Reply:
One downside of using `testing.M` is that it's run for every test, even if the tests don't need it, which can slow down development work.

How much does running `TestMain` in each test add to the test duration? I.e. is it worth adding the code from `TestMain` to `setupTestHelper`?

Reply:
I like this! We could certainly make this a `sync.Once`-style implementation instead of `TestMain`, then any tests that don't use this won't incur the cost, while we still get memoization. Let me explore this, even though I expect 99% of tests to transition to this model at least in this repository, as it's also tidier.

Reply:
>  and I'm not sure it is worth it to have the tests run faster.

After mocked tests, slow tests are my next biggest impediment. Mocked tests are lightning fast, but often meaningless. E2E tests are molasses slow, but arguably most correct. The sweet spot is something that approximates E2E with the speed of mocked tests which is where we land with some of these tradeoffs. In this case, clearing the database is an order of magnitude faster than a fresh database.

The test helper paradigm here (and in the server) automates this for the most part, with each new `Test*` getting a fresh set of tables related to the plugin. We spin up new teams, channels and users all over the place without clearing them between tests, but I haven't yet found this to be problematic (unless one starts hard-coding usernames or something).

I do need to fix one thing in the `testHelper` setup to ensure a fresh instance also clears the database, but I'm happy to own this going forward if we find this more painful than I predict.

Reply:
If the tables are cleared automatically I have no issue.